### PR TITLE
Implemented cameras to follow the adventurer.

### PR DESCRIPTION
**Pull request by ehgoodenough**
### Link

https://mocsarcade.github.io/enchiridion/pr2
### Screenshot

pr2

### CODE REVIEW

File: source/index.js [hunk 1]
```diff
@@ -47,9 +48,11 @@ class Mount extends React.Component {
...
             return (
                 <AspectRatioFrame width={320} height={180}>
-                    {this.state.game.entities.map((entity, index) => {
-                        return <Entity data={entity} key={index}/>
-                    })}
+                    <Camera data={this.state.game.camera}>
+                        {this.state.game.entities.map((entity, index) => {
+                            return <Entity data={entity} key={index}/>
+                        })}
+                    </Camera>
```
Comment (ehgoodenough) [inline source/index.js:55]: The camera is literally just a div that contains all the stuff in the frame.

File: source/scripts/model/Adventurer.js [hunk 2]
```diff
@@ -51,6 +51,11 @@ class Adventurer {
         
         this.position.x += movement.x
         this.position.y += movement.y
+        
+        if(this.game != undefined
+        && this.game.camera != undefined) {
+            this.game.camera.center(this.position)
+        }
```
Comment (ehgoodenough) [inline source/scripts/model/Adventurer.js:58]: Whenever accessing the game object, you don't know what's inside it, so be careful to check, so we don't get null errors.

File: source/scripts/model/Camera.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        var x = Math.min(x0, x1)
+        var y = Math.min(y0, y1)
+        var w = Math.abs(x0 - x1)
+        var h = Math.abs(y0 - y1)
+        
+        this.zoom = Math.min(this.width / w, this.height /  h)
+        this.position.x = x - (((this.width / this.zoom) - w) / 2)
+        this.position.y = y - (((this.height / this.zoom) - h) / 2)
+    }
+}
```
Comment (ehgoodenough) [inline source/scripts/model/Camera.js:31]: This function will remain unused for now. But we still support it. :]

File: source/scripts/model/Dungeon.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        this.rooms.push(new Room({
+            position: {x: -3, y: -3},
+            width: 5, height: 3,
+            color: "#555"
+        }))
+        this.rooms.push(new Room({
+            position: {x: 5, y: 7},
+            width: 3, height: 3,
+            color: "#777"
+        }))
```
Comment (ehgoodenough) [inline source/scripts/model/Dungeon.js:45]: I've just hardcoded a ton of rooms. Eventually, we'll replace all of this with an algorithm that'll generate the rooms of the dungeon programatically.

**Comment by ehgoodenough** [comment]
The construction site across the street from me is very noisy right now.

[selfie-0](https://github.com/thieman/github-selfies/)

**Comment by ehgoodenough** [comment]
These were the experimental builds I made to AB test whether a centering or zooming camera would be better.

https://mocsarcade.github.io/enchiridion/pr2a (center the adventurer)
https://mocsarcade.github.io/enchiridion/pr2b (zooms out on the room the adventurer is in)

**Comment by johndyoung** [comment]
dig the div with transform as the camera

**Pull request merged**